src/gf.c

```diff
@@ -281,11 +281,9 @@ jl_code_info_t *jl_type_infer(jl_method_instance_t *mi, size_t world, int force)
         return NULL;
     jl_task_t *ct = jl_current_task;
     if (ct->reentrant_inference == (uint16_t)-1) {
-        // TODO: We should avoid attempting to re-inter inference here at all
-        // and turn on this warning, but that requires further refactoring
-        // of the precompile code, so for now just catch that case here.
-        //jl_printf(JL_STDERR, "ERROR: Attempted to enter inference while writing out image.");
-        return NULL;
+        // We must avoid attempting to re-enter inference here
+        assert(0 && "attempted to enter inference while writing out image");
+        abort();
```

Comment:
aborting here in non-debug mode seems a bit aggressive. Generally the image may still be ok even if this fails, and it seems unfriendly to the user to not try and continue.

Reply:
There is not really any points in this code that would be calling back into julia now, so I had considered deleting this special flag, but I did want to make it hard for someone to introduce the same mistake again. In most later places, the image might also not be complete, which also seemed more aggravating to encounter for the user than a clear abort and backtrace from the source of the problem